android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/translate/GeckoTranslateSessionDelegate.kt

```diff
@@ -0,0 +1,79 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package mozilla.components.browser.engine.gecko.translate
+
+import mozilla.components.browser.engine.gecko.GeckoEngineSession
+import mozilla.components.concept.engine.translate.DetectedLanguages
+import mozilla.components.concept.engine.translate.GeckoTranslationState
+import mozilla.components.concept.engine.translate.TranslationPair
+import org.mozilla.geckoview.GeckoSession
+import org.mozilla.geckoview.TranslationsController
+
+internal class GeckoTranslateSessionDelegate(
+    private val engineSession: GeckoEngineSession,
+) : TranslationsController.SessionTranslation.Delegate {
+
+    /**
+     * This delegate function is triggered when requesting a translation on the page is likely.
+     *
+     * The criteria is that the page is in a different language than the user's known languages and
+     * that the page is translatable (a model is available).
+     *
+     * @param session The session that this delegate event corresponds to.
+     */
+    override fun onExpectedTranslate(session: GeckoSession) {
+        engineSession.notifyObservers {
+            onTranslateExpected()
+        }
+    }
+
+    /**
+     * This delegate function is triggered when the app should offer a translation.
+     *
+     * The criteria is that the translation is likely and it is the user's first visit to the host site.
+     *
+     * @param session The session that this delegate event corresponds to.
+     */
+    override fun onOfferTranslate(session: GeckoSession) {
+        engineSession.notifyObservers {
+            onTranslateOffer()
+        }
+    }
+
+    /**
+     * This delegate function is triggered when the state of the translation or translation options
+     * for the page has changed. State changes usually occur on navigation or if a translation
+     * action was requested, such as translating or restoring to the original page.
+     *
+     * This provides the Gecko translations toolkit state and information for the page.
+     *
+     * @param session The session that this delegate event corresponds to.
+     * @param geckoTranslationsState The toolkit reported translations state. Not to be confused
+     * with the AC browser translation state.
+     */
+    override fun onTranslationStateChange(
+        session: GeckoSession,
+        geckoTranslationsState: TranslationsController.SessionTranslation.TranslationState?,
+    ) {
+        var detectedLanguages = DetectedLanguages(
+            geckoTranslationsState?.detectedLanguages?.docLangTag,
+            geckoTranslationsState?.detectedLanguages?.isDocLangTagSupported,
+            geckoTranslationsState?.detectedLanguages?.userLangTag,
+        )
+        var pair = TranslationPair(
+            geckoTranslationsState?.requestedTranslationPair?.fromLanguage,
+            geckoTranslationsState?.requestedTranslationPair?.toLanguage,
+        )
```

Comment:
Could `docLangTag`, `isDocLangTagSupported`, `userLangTag`, and `toLanguage` just be part of `TranslationState` class?  Could you help me to clarify why they need to be part of separate classes.

Reply:
Possible, but would need a little more discussion we'd like to go that route! (More info in this [thread](https://github.com/mozilla-mobile/firefox-android/pull/4465#discussion_r1399765495 ).) Mostly just a decision if we should hide that those components have the same lifecycle or not.